os/drivers/audio/alc5658.c

```diff
@@ -974,8 +974,31 @@ static int alc5658_ioctl(FAR struct audio_lowerhalf_s *dev, int cmd, unsigned lo
 	switch (cmd) {
 
 	case AUDIOIOC_PREPARE: {
-		/* TODO: Need to be implemented later */
-		audvdbg("AUDIOIOC_PREPARE: No Action Would be taken now \n");
+		/* This is to aid audio prepare, this would stop dma, i2s and
+		cleanup all queues. Idea behind is to use this in xrun cases to stop
+		followed by (restart) but restart is not part of this */
+		audvdbg("AUDIOIOC_PREPARE: Stop dma,i2s and cleanup qs\n");
+
+		/* Take semaphore */
+		alc5658_takesem(&priv->devsem);
+
+		/* Stop dma, i2s */
+		I2S_STOP(priv->i2s);
+
+		/* Switch alc status */
+		priv->running = false;
+
+		/* Clean up alc queue */
+		sq_entry_t *tmp = NULL;
+		for (tmp = (sq_entry_t *)sq_peek(&priv->pendq); tmp; tmp = sq_next(tmp)) {
+			sq_rem(tmp, &priv->pendq);
```

Comment:
This will cause problem if pcm_prepare is called in pcm_start in tinyalsa.
- i will remove pcm_prepare call from pcm_start in tinyalsa.  
- then pcm_prepare will be called onlly for underrun cases..
- if user calls pcm_prepare during normal operation, it is his responsibility to enqueue new buffers..

Reply:
@Taejun-Kwon  A clear definition on pcm_prepare and when it has to be used has to be defined. If we follow tinyalsa, it has to be called at pcm_start.

Reply:
As discussed offline, "prepare" will not do any buffer handling. 
Please remove buffer handling code.